File: src/data/courses/dsa/trees/path-relationship-problems/kth-ancestor/code/python/kth_ancestor.py
# ============================================================
# Kth Ancestor
# ============================================================
class Solution:
    def findKthAncestor(self, root, target: int, k: int) -> int:
        path = []
        def find_path(node):
            if not node: return False
            path.append(node)
            if node.val == target: return True
            if find_path(node.left) or find_path(node.right): return True
            path.pop()
            return False
        find_path(root)
        idx = len(path) - 1 - k
        return path[idx].val if idx >= 0 else -1

File: src/data/courses/dsa/trees/path-relationship-problems/kth-ancestor/code/java/KthAncestor.java
// ============================================================
// Kth Ancestor
// ============================================================
import java.util.*;
class Solution {
    public int findKthAncestor(TreeNode root, int target, int k) {
        List<TreeNode> path = new ArrayList<>();
        findPath(root, target, path);
        int idx = path.size() - 1 - k;
        return idx >= 0 ? path.get(idx).val : -1;
    }
    boolean findPath(TreeNode node, int target, List<TreeNode> path) {
        if (node == null) return false;
        path.add(node);
        if (node.val == target) return true;
        if (findPath(node.left, target, path) || findPath(node.right, target, path)) return true;
        path.remove(path.size()-1);
        return false;
    }
}
